### ISSUE TITLE
Supports multi-machine training

### ISSUE THREAD

**Issue by dongyibo** [sgl-project/SpecForge#373]
Hello, does SpecForge support multi-machine training mode? 
I've been training with a single machine with 8 GPUs, but now the training data volume has increased significantly. I'd like to use two single machines with 8 GPUs each, totaling 16 GPUs for training. Is this supported currently?

**Comment by FrankLeeeee** [comment]
Hi yes, it is surely supported. For example, you can launch `torchrun` on two nodes with `--node-rank` and `--nnodes`.

**Comment by jiapingW** [comment]
I have trained eagle3 model online with 128 X A100. It works well.

**Comment by pengyao96** [comment]
> I have trained eagle3 model online with 128 X A100. It works well.

can you provide a running script?

**Comment by jiapingW** [comment]
```NPROC_PER_NODE=8

export NNODES=${WORLD_SIZE}
export NODE_RANK=${RANK}
export MASTER_ADDR=${MASTER_ADDR}
export MASTER_PORT=${MASTER_PORT}


torchrun \
    --nnodes $NNODES \
    --node_rank $NODE_RANK \
    --master_addr $MASTER_ADDR \
    --master_port $MASTER_PORT \
    --nproc_per_node $NPROC_PER_NODE \
    $ROOT_DIR/scripts/train_eagle3.py \
    --target-model-path Qwen/Qwen3-8B \
    --draft-model-config $ROOT_DIR/configs/qwen3-8b-eagle3.json \
    --train-data-path all.jsonl \
    --output-dir outputs/qwen3 \
    --num-epochs 1 \
    --batch-size 1 \
    --learning-rate 1e-4 \
    --max-length 4096 \
    --chat-template qwen \
    --cache-dir $ROOT_DIR/cache \
    --attention-backend flex_attention \
    --save-interval 2500 \
    --tp-size 1 \
    --report-to tensorboard \
    --target-model-backend custom \
    --dist-timeout 6000 \
```
> > I have trained eagle3 model online with 128 X A100. It works well.
> 
> can you provide a running script?

**Comment by pengyao96** [comment]
> --nnodes

thanks for you reply!  
I have another problem, how to use multi-node to extract offline feature? i.e. run prepare_hidden_states.py 
can you provide a running script like this...
